### PR TITLE
Mapping fixes - fixes #1005

### LINKED ISSUE #1005: Component Mapping Broken!
Look in the Tests "SelectTests.SelectComplex1 - 3", the Middle and Second Name are always "null". They are not selected from the database

### PR DESCRIPTION
**Pull request by jogibear9988**

### CODE REVIEW

File: Source/LinqToDB/Metadata/MetadataReader.cs [hunk 1]
```diff
@@ -20,10 +23,16 @@ public class MetadataReader : IMetadataReader
...
+			_readers = new List<IMetadataReader>(readers) ?? throw new ArgumentNullException(nameof(readers));
 		}
 
-		readonly IMetadataReader[] _readers;
+		IList<IMetadataReader> _readers;
+
+		internal void AddReader(IMetadataReader reader)
+		{
+			// creation of new list is cheaper than lock on each method call
+			_readers = new[] { reader }.Concat(_readers).ToList();
```
Comment (sdanyliv) [inline Source/LinqToDB/Metadata/MetadataReader.cs]: So, why not Array here?

File: Source/LinqToDB/Mapping/EntityDescriptor.cs [hunk 2]
```diff
@@ -130,21 +130,27 @@ void Init(MappingSchema mappingSchema)
 					continue;
 				}
 
-				var ca = mappingSchema.GetAttribute<ColumnAttribute>(TypeAccessor.Type, member.MemberInfo, attr => attr.Configuration);
+				var cas = mappingSchema.GetAttributes<ColumnAttribute>(TypeAccessor.Type, member.MemberInfo, attr => attr.Configuration);
 
-				if (ca != null)
+				if (cas != null && cas.Length > 0)
```
Comment (sdanyliv) [inline Source/LinqToDB/Mapping/EntityDescriptor.cs]: Just for clarification, why several column attributes are needed?

File: Source/LinqToDB/Metadata/MetadataReader.cs (hunk 1 above)
Comment (jogibear9988) [reply Source/LinqToDB/Metadata/MetadataReader.cs]: @MaceWindu can you look? or should I change?

File: Source/LinqToDB/Mapping/EntityDescriptor.cs (hunk 2 above)
Comment (jogibear9988) [reply Source/LinqToDB/Mapping/EntityDescriptor.cs]: look here: https://github.com/linq2db/linq2db/blob/b02ddd52cb496c7a1252fd49befa57d92f3eb75b/Tests/Model/ComplexPerson.cs#L11

Comment (jogibear9988) [reply Source/LinqToDB/Mapping/EntityDescriptor.cs]: The tests SelectComplex1-3 never returned the correct db value

**Comment by jogibear9988** [comment]
@sdanyliv 
I now see my error.... The Column Attribute is on the class not the column.

The issue is, for ComplexTest3, the FluentMetadatareader does return the ColumnAttribute not for the class it returns them for the property...

I will fix that...

**Comment by jogibear9988** [comment]
Atm I've no clue how to fix. Should a mapping for Complex Properties be on the class or the Property?

File: Source/LinqToDB/Metadata/MetadataReader.cs (hunk 1 above)
Comment (jogibear9988) [reply Source/LinqToDB/Metadata/MetadataReader.cs]: https://stackoverflow.com/a/16323412

**Comment by sdanyliv** [comment]
Never use this functionality, i have to look at this closer.

**Comment by jogibear9988** [comment]
I've now fixed it (hopefully). I now Add FluentMappings (if they specify SubProperties) on the class instead of the Property (as done with attribute mappings)

File: Source/LinqToDB/Metadata/MetadataReader.cs (hunk 1 above)
Comment (MaceWindu) [reply Source/LinqToDB/Metadata/MetadataReader.cs]: Yes, I forgot to change it back to array... no list needed here

**Comment by jogibear9988** [comment]
@MaceWindu @sdanyliv I've fixed the ToList into ToArray. Think it's done, can anyone approve?

**Comment by jogibear9988** [comment]
@MaceWindu this one? ;-)

**Pull request merged**